Do not count tests, instead use done_testing()

mbeijen opened the pull request:
this is my first contribution to perl5

Having to count tests is annoying for maintainers.

Also, manually updating count tests is problematic when multiple people
are working on the same code; it causes merge conflicts and recounts.

Sometimes creative workarounds involving BEGIN blocks were used to make
counting the tests a little more bearable. But these tricks are no
longer required.

done_testing() is available since Test::More 0.88 which was released in
2009.

While updating this I've also removed shebang lines if possible - there
were many different shebangs in use, and the files do not have an
execute bit. I've also added 'use strict; use warnings;' if possible.

Some of the tests use 'test.pl' and this is most often not super easy to
change because they use features of 'test.pl' not available in
Test::More; such as fresh_perl_*(); luckily also `test.pl` has
done_testing()!

This PR covers all tests in lib/ that were using Test::More and a plan.

(updated: previously I also made minor other cleanups, which increased the complexity of the PR, probably a bit too much)

jkeenan commented:
> this is my first contribution to perl5
> 

Welcome!  And thanks for taking the time to contribute to Perl 5.


> Having to count tests is annoying for maintainers.
> 
> Also, manually updating count tests is problematic when multiple people are working on the same code; it causes merge conflicts and recounts.
> 
> Sometimes creative workarounds involving BEGIN blocks were used to make counting the tests a little more bearable. But these tricks are no longer required.
> 

I'm with you so far ... but with an important caveat.  You have made changes to **70** files in the core distribution.  That is a lot for reviewers to plow through.  My eyes glaze over at that number of changes.

I feel that it would have been better to discuss this, either on #p5p or on our mailing list, before touching such a large number of files.  It might have been better to structure this as a **series** of smaller pull requests.

> done_testing() is available since Test::More 0.88 which was released in 2009.
> 
> While updating this I've also removed shebang lines if possible - there were many different shebangs in use, and the files do not have an execute bit. I've also added 'use strict; use warnings;' if possible.
> 

This is where I disagree with you.  (Note that I'm speaking for myself here and other P5P committers may disagree with me.)  I believe it's better to focus on one type of clean-up in a given pull request.  For example, clean-up of shebang lines would be done in one pull request or in a series of smaller pull requests.  Adding `use strict` and `use warnings` would similarly be better done in pull requests focused on each of those.

(I recommend this on the basis of past experience.  Adding `strict` and `warnings` to test files has been surprisingly controversial in the past -- and I speak as someone who, in the no-longer-extant Perl 7 project in 2020, tried to get the entire test suite running with `strict` and `warnings` on by default.  We managed to succeed with `strict` but not with `warnings`.)

> Some of the tests use 'test.pl' and this is most often not super easy to change because they use features of 'test.pl' not available in Test::More; such as fresh_perl_*(); luckily also `test.pl` has done_testing()!
> 
> This PR covers most of the files in ext/ and lib/
> 

In particular, revisions to files using `test.pl` should be done separately from those using `Test::More`.

> I've been listening on #p5p on irc.perl.org and noticed a conversation on this topic; I realized that most tests for perl internals in t/ can also be updated to use 'done_testing();'

I'm frequently available on #p5p U.S. Easter time daytimes (as `kid51`).  Please feel free -- no, more than free, encouraged -- to join the conversations there so that we can have real-time back-and-forth on this.

But, once again, thank you for your efforts and we look forward to working with you on this.

Thank you very much.
Jim Keenan

mbeijen committed:
do not count tests, instead use done_testing()

Having to count tests is annoying for maintainers.

Also, manually updating count tests is problematic when multiple people
are working on the same code; it causes merge conflicts and recounts.

done_testing() is available since Test::More 0.88 which was released in
2009.

This commit changes all tests under lib/ that use Test::More and were
planning the number of tests.

Michiel Beijen is now a Perl author

leonerd commented:
On @jkeenan 's point here:

> I'm with you so far ... but with an important caveat. You have made changes to 70 files in the core distribution. That is a lot for reviewers to plow through. My eyes glaze over at that number of changes.

> I feel that it would have been better to discuss this, either on #p5p or on our mailing list, before touching such a large number of files. It might have been better to structure this as a series of smaller pull requests.

This change was infact discussed on IRC - with me :) So I was anticipating seeing it.

Though I would agree with jkeenan's comment that it would be better to split this into a few separate changes. `done_testing()` in one commit, and separately look into the strict+warnings. We have some history of that and it turns out for a bunch of files it isn't as simple as we'd hope to just add those, because of decades of accumulated cruft and bad habits in some files.

leonerd commented:
*Update*: Oh, I see on re-reading that you have in fact done just that. Yes this is looking much better now. I'll press the CI box and see what that makes of things.

leonerd commented:
A couple of failures on the CI run there that you might want to look into

khwilliamson commented:
On 12/3/21 9:34 AM, Paul Evans wrote:
> On @jkeenan <https://github.com/jkeenan> 's point here:
> 
>     I'm with you so far ... but with an important caveat. You have made
>     changes to 70 files in the core distribution. That is a lot for
>     reviewers to plow through. My eyes glaze over at that number of changes.
> 
>     I feel that it would have been better to discuss this, either on
>     #p5p or on our mailing list, before touching such a large number of
>     files. It might have been better to structure this as a series of
>     smaller pull requests.
> 
> This change was infact discussed on IRC - with me :) So I was 
> anticipating seeing it.
> 
> Though I would agree with jkeenan's comment that it would be better to 
> split this into a few separate changes. |done_testing()| in one commit, 
> and separately look into the strict+warnings. We have some history of 
> that and it turns out for a bunch of files it isn't as simple as we'd 
> hope to just add those, because of decades of accumulated cruft and bad 
> habits in some files.
> 
> —
>

There are people who think that done_testing is not as bomb-proof as a 
count, and have hence opposed changing tests.

I have never understood their position.

But if you're such a person, now is the time to speak up, with your reasons.

mbeijen commented:
I updated the PR to reduce the scope; the CI fail should also be gone now.

leonerd commented:
Oddly, some of the CI workflows are failing the authors test; but others are not. The failure is:

```
 Michiel W. Beijen  <michiel.beijen@gmail.com> not found in AUTHORS
porting/authors.t .................................................... 
Failed 1/1 subtests 
```

I'm entirely unsure why that would only affect some of them.

jkeenan commented:
> Oddly, some of the CI workflows are failing the authors test; but others are not. The failure is:
> 
> ```
>  Michiel W. Beijen  <michiel.beijen@gmail.com> not found in AUTHORS
> porting/authors.t .................................................... 
> Failed 1/1 subtests 
> ```
> 
> I'm entirely unsure why that would only affect some of them.

I'll take this ticket for the purpose of (a) fixing up any porting tests; (b) pausing a few days per the concerns @khwilliamson raised above; and (c) merging it into blead.

jkeenan commented:
@mbeijen, you've got 2 stray lines in `lib/sort.` which are causing these warnings to be thrown:
```
Useless use of addition (+) in void context at ./lib/sort.t line 29.
Argument "*main::3" isn't numeric in addition (+) at ./lib/sort.t line 29.
```
Can you fix?

Thank you very much.
Jim Keenan

mbeijen commented:
Thanks for noticing! I fixed `lib/sort.t`.
